src/components/contact-govuk-one-login/contact-govuk-one-login-controller.ts

```diff
@@ -9,114 +9,174 @@ import {
   supportWebchatContact,
 } from "../../config";
 import { generateReferenceCode } from "../../utils/referenceCode";
+import { eventService } from "./event-service";
+import { AuditEvent, Extensions, Platform, User } from "./types";
 
 const CONTACT_ONE_LOGIN_TEMPLATE = "contact-govuk-one-login/index.njk";
+const MISSING_SESSION_VALUE_SPECIAL_CASE : string = "";
 
-export function contactGet(req: Request, res: Response): void {
-  const isAuthenticated = req.session?.user?.isAuthenticated;
-  let isLoggedOut = req.cookies?.lo;
-  if (typeof isLoggedOut === "string") {
-    isLoggedOut = JSON.parse(isLoggedOut);
+export function contactGet(req: Request, res: Response, ): void {
+  logger.info(req.session, "The session contains:")
+  updateSessionFromQueryParams(req.session, req.query);
+  const audit_event = buildAuditEvent(req);
+  logUserVisitsContactPage(audit_event);
+  sendUserVisitsContactPageAuditEvent(audit_event);
+  render(req, res);
+}
+
+const updateSessionFromQueryParams = (session: any, queryParams: any): void => {
+  if (isValidUrl(queryParams.fromURL as string)) {
+    session.fromURL = queryParams.fromURL;
+  } else {
+    logger.error("fromURL in request query for contact-govuk-one-login page did not pass validation");
   }
 
-  const { originalUrl, language, protocol, hostname } = req;
-  const baseUrl = protocol + "://" + hostname;
+  copySafeQueryParamToSession(session, queryParams, 'theme');
+  copySafeQueryParamToSession(session, queryParams, 'appSessionId');
+  copySafeQueryParamToSession(session, queryParams, 'appErrorCode');
 
-  const referenceCode = req.session.referenceCode
-    ? req.session.referenceCode
-    : generateReferenceCode();
-  req.session.referenceCode = referenceCode;
+  if (!session.referenceCode) {
+    session.referenceCode = generateReferenceCode();
+  }
+  logger.info("completed updating session");
+}
 
-  const contactEmailServiceUrl =
-    buildContactEmailServiceUrlAndSaveDataToSession(req).toString();
+const copySafeQueryParamToSession = (session: any, queryParams: any, paramName: string) => {
+  if (queryParams[paramName] && isSafeString(queryParams[paramName] as string)) {
+    session[paramName] = queryParams[paramName];
+  } else {
+    logger.error(`${paramName} in request query for contact-govuk-one-login page did not pass validation`);
+  }
+};
 
-  logContactDataFromSession(req);
+const buildAuditEvent = (req: Request): AuditEvent => {
+  const session: any = req.session;
+  let sessionId: string;
 
-  const data = {
-    contactWebchatEnabled: supportWebchatContact(),
-    contactPhoneEnabled: supportPhoneContact(),
-    showContactGuidance: showContactGuidance(),
-    showSignOut: isAuthenticated && !isLoggedOut,
-    referenceCode,
-    contactEmailServiceUrl: contactEmailServiceUrl,
-    webchatSource: getWebchatUrl(),
-    currentUrl: originalUrl,
-    baseUrl,
-    language,
+  if (userHasSignedIntoHomeRelyingParty(session)) {
```

Comment:
This can be simplified to
`const sessionId = session.user?.sessionId ?? MISSING_SESSION_VALUE_SPECIAL_CASE;`

Reply:
I want to keep the clarifying function calls as they provide information about what the check actually means to the user.